### PR TITLE
fix(challenge): use channel to trigger challenge worker instantly

### PR DESCRIPTION
**Pull request by cnlangzi**
### Changes
-

### Fixes
-  use channel to trigger next challenge worker instantly. 

### Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/blobber/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

### Associated PRs (Link as appropriate):
- 0chain:

### CODE REVIEW

File: code/go/0chain.net/blobbercore/challenge/challenge.go [hunk 1]
```diff
@@ -52,12 +54,11 @@ func syncOpenChallenges(ctx context.Context) {
 	}
 
 	for _, challengeObj := range blobberChallenges.Challenges {
-
 		if challengeObj == nil || challengeObj.ChallengeID == "" {
 			logging.Logger.Info("[challenge]open: No challenge entity from the challenge map")
 			continue
 		}
-
+		log.Println("challenge: sync", challengeObj.ChallengeID)
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/challenge/challenge.go]: Let's use logging

File: code/go/0chain.net/blobbercore/challenge/challenge.go [hunk 2]
```diff
@@ -169,6 +185,7 @@ func commitProcessed(ctx context.Context) {
 		for _, challenge := range challenges {
 			c := challenge
 			swg.Add()
+			log.Println("challenge: commit ", time.Now())
```
Comment (peterlimg) [inline code/go/0chain.net/blobbercore/challenge/challenge.go]: Use logging

**Review by peterlimg — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/blobbercore/challenge/worker.go [hunk 3]
```diff
@@ -30,6 +37,8 @@ func startProcessAccepted(ctx context.Context) {
 		select {
 		case <-ctx.Done():
 			return
+		case next := <-waitToProcess:
```
Comment (ssardana08) [inline code/go/0chain.net/blobbercore/challenge/worker.go:40]: do we still needs to be called from `processAccepted` ?

File: code/go/0chain.net/blobbercore/challenge/worker.go [hunk 4]
```diff
@@ -14,11 +14,18 @@ func SetupWorkers(ctx context.Context) {
...
+	waitToCommit  = make(chan *ChallengeEntity)
+)
+
 func startCommitProcessed(ctx context.Context) {
 	for {
 		select {
 		case <-ctx.Done():
 			return
+		case next := <-waitToCommit:
+			commitChallenge(next)
```
Comment (ssardana08) [inline code/go/0chain.net/blobbercore/challenge/worker.go:28]: do we still needs to be called from `commitProcessed` ?

**Review by ssardana08 — CHANGES_REQUESTED** [review]

File: code/go/0chain.net/blobbercore/challenge/worker.go (hunk 3 above)
Comment (cnlangzi) [reply code/go/0chain.net/blobbercore/challenge/worker.go:40]: @sum-it08  Yes. it is keep challenge will be lost to process even blobber crashes and restarts.

File: code/go/0chain.net/blobbercore/challenge/worker.go (hunk 4 above)
Comment (cnlangzi) [reply code/go/0chain.net/blobbercore/challenge/worker.go:28]: @sum-it08  Yes. it is keep challenge will be lost to commit even blobber crashes and restarts.

File: code/go/0chain.net/blobbercore/challenge/challenge.go (hunk 1 above)
Comment (cnlangzi) [reply code/go/0chain.net/blobbercore/challenge/challenge.go]: fixed

File: code/go/0chain.net/blobbercore/challenge/challenge.go (hunk 2 above)
Comment (cnlangzi) [reply code/go/0chain.net/blobbercore/challenge/challenge.go]: fixed

File: code/go/0chain.net/blobbercore/challenge/challenge.go [hunk 5]
```diff
@@ -131,7 +135,11 @@ func processAccepted(ctx context.Context) {
 }
 
 func validateChallenge(swg *sizedwaitgroup.SizedWaitGroup, challengeObj *ChallengeEntity) {
```
Comment (ssardana08) [inline code/go/0chain.net/blobbercore/challenge/challenge.go:137]: for this to be called we should just be playing with `challenge_response.frequency`
make it shorter if we need it to be called frequently

**Review by ssardana08 — COMMENTED** [review]
@cnlangzi @peterlimg I don't think this PR would be adding much value, but making challenge process more complex
We have a single config `challenge_response.frequency` which triggers process, validate, commit delay of challenge. We can just make that value small to process challenges faster. That's all. We don't need these new changes to make the challenge process faster

**Comment by ssardana08** [comment]
changes to `challenge_response.frequency` should be able to make the challenge process fast.
We can close this for now

**Pull request closed without merging**